File: src/main/java/io/github/jsnimda/common/gui/ScrollableContainerWidget.java
package io.github.jsnimda.common.gui;

import java.util.Optional;

import com.mojang.blaze3d.platform.GlStateManager;

import org.lwjgl.opengl.GL11;

import net.minecraft.client.gui.Drawable;
import net.minecraft.client.gui.DrawableHelper;
import net.minecraft.client.gui.Element;

public class ScrollableContainerWidget extends DrawableHelper implements Drawable, Element {

  private int contentHeight = 0;
  private int scrollY = 0;
  private int x;
  private int y;
  private int width;
  private int height;

  public boolean renderBorder = true;
  public int borderColor = COLOR_BORDER;
  public int padding = 3;
  private int scrollbarWidth = 6;

  private Optional<IContentRenderer> contentRenderer = Optional.empty();

  public void setContentRenderer(IContentRenderer contentRenderer) {
    this.contentRenderer = Optional.ofNullable(contentRenderer);
  }

  public void setBounds(int x, int y, int width, int height) {
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
    setScrollY(scrollY);
  }
  
  //#region bounds getter
  public int getX() {
    return x;
  }
  public int getY() {
    return y;
  }
  public int getWidth() {
    return width;
  }
  public int getHeight() {
    return height;
  }
  //#endregion
  
  //#region viewport getter
  public int getViewportX() {
    return x + padding;
  }
  public int getViewportY() {
    return y + padding;
  }
  public int getViewportWidth() {
    return width - scrollbarWidth - padding * 2;
  }
  public int getViewportHeight() {
    return height - padding * 2;
  }
  //#endregion
  
  public int getScrollYMax() {
    return Math.max(0, contentHeight - getViewportHeight());
  }
  public int getScrollbarThumbHeight() {
    int vh = getViewportHeight();
    if (getScrollYMax() == 0) return vh;
    return vh * vh / contentHeight;
  }
  public int getScrollbarYOffsetMax() {
    return getViewportHeight() - getScrollbarThumbHeight();
  }
  public int getScrollbarYOffset() {
    return scrollYToScrollbarYOffset(scrollY);
  }

  private int scrollbarYOffsetToScrollY(int scrollbarY) {
    if (getScrollbarYOffsetMax() == 0) return 0;
    return map(getScrollbarYOffsetMax(), getScrollYMax(), scrollbarY);
  }
  private int scrollYToScrollbarYOffset(int scrollY) {
    if (getScrollYMax() == 0) return 0;
    return map(getScrollYMax(), getScrollbarYOffsetMax(), scrollY);
  }
  private int map(int inputMax, int outputMax, int input) { // min = 0
    return (int)Math.round(1.0 * input * outputMax / inputMax);
  }

  public void setScrollY(int newScrollY) {
    int oldScrollY = scrollY;
    if (newScrollY < 0) scrollY = 0;
    else scrollY = newScrollY > getScrollYMax() ? getScrollYMax() : newScrollY;
    if (oldScrollY != newScrollY) {
      scrollYChanged(newScrollY, oldScrollY);
    }
  }
  private void scrollYChanged(int newScrollY, int oldScrollY) {
  }
  public void setContentHeight(int contentHeight) {
    this.contentHeight = contentHeight;
    setScrollY(scrollY);
  }

  public int getContentHeight() {
    return contentHeight;
  }
  public int getScrollY() {
    return scrollY;
  }

  // private static final int COLOR_WHITE              = 0xFFFFFFFF;
  private static final int COLOR_BORDER             = 0xFF999999;
  private static final int COLOR_SCROLLBAR_BG       = 0x80000000; // 0xFF000000; ref: EntryListWidget.render
  private static final int COLOR_SCROLLBAR_SHADOW   = 0xFF989898; // 0xFF808080;
  private static final int COLOR_SCROLLBAR          = 0xFFD8D8D8; // 0xFFC0C0C0;
  private static final int COLOR_SCROLLBAR_HOVER_SHADOW   = 0xFFC0C0C0;
  private static final int COLOR_SCROLLBAR_HOVER          = 0xFFFFFFFF;
  @Override
  public void render(int mouseX, int mouseY, float partialTicks) {
    if (renderBorder) {
      VHLine.outline(x, y, x + width, y + height, borderColor);
    }

    // render scrollbar, ref: EntryListWidget.render
    if (isScrollbarVisible()) {
      int x1 = getViewportX() + getViewportWidth();
      int y1 = getViewportY();
      int x2 = x1 + scrollbarWidth;
      int y2 = y1 + getViewportHeight();
      int sy1 = y1 + getScrollbarYOffset();
      int sy2 = sy1 + getScrollbarThumbHeight();
      fill(x1, y1, x2, y2, COLOR_SCROLLBAR_BG);
      boolean hover = VHLine.contains(x1, sy1, x2, sy2, mouseX, mouseY) || scrolling;
      fill(x1, sy1, x2, sy2, hover ? COLOR_SCROLLBAR_HOVER_SHADOW : COLOR_SCROLLBAR_SHADOW);
      fill(x1, sy1, x2 - 1, sy2 - 1, hover ? COLOR_SCROLLBAR_HOVER : COLOR_SCROLLBAR);
      // fill(x1, sy1, x2, sy2, hover ? COLOR_SCROLLBAR_HOVER : COLOR_SCROLLBAR);
    }

    // render content
    if (contentRenderer.isPresent()) {
      GlStateManager.pushMatrix();
      GlStateManager.translatef(0, 0, -400.0F); // ref: AdvancementsScreen widget
      GlStateManager.enableDepthTest();
      GlStateManager.depthFunc(GL11.GL_GEQUAL); // 518

      // draw mask
      GlStateManager.disableAlphaTest();
      int vx = getViewportX();
      int vy = getViewportY();
      int vw = getViewportWidth();
      int vh = getViewportHeight();
      fill(vx, vy, vx + vw, vy + vh, 0);
      GlStateManager.enableAlphaTest();

      // exit draw mask
      GlStateManager.depthFunc(GL11.GL_LEQUAL); // 515, return default
      contentRenderer.get().render(mouseX, mouseY, partialTicks, vx, vy, vw, vh, scrollY);
  
      GlStateManager.popMatrix();
      GlStateManager.disableDepthTest();
    }

  }

  public boolean isScrollbarVisible() {
    return getScrollYMax() > 0;
  }

  private boolean hoverScrollbarThumb(int mouseX, int mouseY) {
    int x1 = getViewportX() + getViewportWidth();
    int y1 = getViewportY();
    int x2 = x1 + scrollbarWidth;
    int sy1 = y1 + getScrollbarYOffset();
    int sy2 = sy1 + getScrollbarThumbHeight();
    return VHLine.contains(x1, sy1, x2, sy2, mouseX, mouseY);
  }
  private boolean hoverScrollbar(int mouseX, int mouseY) {
    int x1 = getViewportX() + getViewportWidth();
    int y1 = getViewportY();
    int x2 = x1 + scrollbarWidth;
    int y2 = y1 + getViewportHeight();
    return VHLine.contains(x1, y1, x2, y2, mouseX, mouseY);
  }

  // scrolling logic / ui events
  
  private boolean scrolling = false;
  private double scrollingInitMouseY;
  private int scrollingInitScrollbarYOffset;
  @Override
  public boolean mouseClicked(double d, double e, int i) {
    if (isScrollbarVisible()) {
      scrolling = hoverScrollbar((int)d,(int)e);
      if (i == 0 && scrolling) { // ref: vscode scrollbar mechanic
        if (!hoverScrollbarThumb((int)d,(int)e)) {
          int y1 = getViewportY();
          // e = y1 + yoffset + sh/2
          int newYOffset = (int)e - y1 - getScrollbarThumbHeight() / 2;
          setScrollY(scrollbarYOffsetToScrollY(newYOffset));
        }
        scrollingInitMouseY = e;
        scrollingInitScrollbarYOffset = getScrollbarYOffset();
        return true;
      }
    }
    return false;
  }

  @Override
  public boolean mouseReleased(double d, double e, int i) {
    scrolling = false;
    return false;
  }

  @Override
  public boolean mouseDragged(double mouseX, double mouseY, int button, double dx, double dy) {
    if (button == 0 && scrolling) {
      double shiftY = mouseY - scrollingInitMouseY;
      setScrollY(scrollbarYOffsetToScrollY(scrollingInitScrollbarYOffset + (int)shiftY));
      return true;
    }
    return false;
  }

  @Override
  public boolean mouseScrolled(double d, double e, double f) { // f = 1 or -1
    if (isScrollbarVisible()) {
      setScrollY(scrollY - (int) (f * 20));
    }
    return true;
  }

  @Override
  public boolean isMouseOver(double d, double e) {
    return VHLine.contains(x, y, x + width, y + height, (int)d, (int)e);
  }

  public boolean isMouseOverViewport(double d, double e) {
    int x1 = getViewportX();
    int y1 = getViewportY();
    int x2 = x1 + getViewportWidth();
    int y2 = y1 + getViewportHeight();
    return VHLine.contains(x1, y1, x2, y2, (int)d, (int)e);
  }

  public interface IContentRenderer {
    void render(int mouseX, int mouseY, float partialTicks, int viewportX, int viewportY, int viewportWidth, int viewportHeight, int scrollY);
  }
}

File: src/main/java/org/anti_ad/mc/common/gui/widgets/ScrollableContainerWidget.kt
package org.anti_ad.mc.common.gui.widgets

import org.anti_ad.mc.common.gui.widget.AnchorStyles
import org.anti_ad.mc.common.gui.widget.Overflow.HIDDEN
import org.anti_ad.mc.common.math2d.Rectangle
import org.anti_ad.mc.common.vanilla.render.rDrawOutline
import org.anti_ad.mc.common.vanilla.render.rFillRect
import kotlin.math.roundToInt

// private static final int COLOR_WHITE              = 0xFFFFFFFF;
private const val COLOR_BORDER = -0x666667
private const val COLOR_SCROLLBAR_BG = -0x80000000 // 0xFF000000; ref: EntryListWidget.render
private const val COLOR_SCROLLBAR_SHADOW = -0x676768 // 0xFF808080;
private const val COLOR_SCROLLBAR = -0x272728 // 0xFFC0C0C0;
private const val COLOR_SCROLLBAR_HOVER_SHADOW = -0x3f3f40
private const val COLOR_SCROLLBAR_HOVER = -0x1

class ScrollableContainerWidget : Widget() {

  private val scrollbarWidth = 6
  private val padding = 3

  val viewport = Widget().apply {
    overflow = HIDDEN
    anchor = AnchorStyles.all
    this@ScrollableContainerWidget.addChild(this)
    top = padding
    left = padding
    bottom = padding
    right = padding + scrollbarWidth
  }

  private val _contentContainer = object : Widget() {
    override fun render(mouseX: Int, mouseY: Int, partialTicks: Float) {
      contentCustomRenderer(mouseX, mouseY, partialTicks)
    }

    fun superRender(mouseX: Int, mouseY: Int, partialTicks: Float) {
      super.render(mouseX, mouseY, partialTicks)
    }
  }.apply {
    anchor = AnchorStyles.noBottom
    viewport.addChild(this)
    top = 0
    left = 0
    right = 0
    sizeChanged += {
      scrollY = scrollY // update scrollY
    }
  }

  val contentContainer
    get() = _contentContainer
  val contentCustomRendererDefault: Widget.(Int, Int, Float) -> Unit =
    { mouseX: Int, mouseY: Int, partialTicks: Float ->
      _contentContainer.superRender(mouseX, mouseY, partialTicks)
    }
  var contentCustomRenderer = contentCustomRendererDefault

  var contentHeight: Int
    get() = contentContainer.height
    set(value) {
      contentContainer.height = value
    }
  var scrollY: Int
    get() = -contentContainer.top
    set(value) {
      contentContainer.top = -value.coerceIn(0, scrollYMax)
    }
  val scrollYMax: Int
    get() = (contentHeight - viewport.height).coerceAtLeast(0)

  var renderBorder = true
  var borderColor = COLOR_BORDER

  //region Scrollbar

  val scrollbar = Scrollbar() // not using "= object { ... }" for public accessing

  inner class Scrollbar {
    val visible: Boolean
      get() = scrollYMax > 0
    val trackHeight: Int
      get() = viewport.height
    val thumbHeight: Int
      get() {
        return if (visible) trackHeight * trackHeight / contentHeight else trackHeight
      }
    val yMax: Int
      get() = trackHeight - thumbHeight
    var y: Int
      get() = toY(scrollY)
      set(value) {
        scrollY = toScrollY(value)
      }

    fun toScrollY(y: Int): Int = map(yMax, scrollYMax, y)
    fun toY(scrollY: Int): Int = map(scrollYMax, yMax, scrollY)

    val trackAbsoluteBounds: Rectangle
      get() = viewport.absoluteBounds.run { copy(x = right, width = scrollbarWidth) }
    val thumbAbsoluteBounds: Rectangle
      get() = trackAbsoluteBounds.let { it.copy(y = it.y + y, height = thumbHeight) }
  }

  private fun map(inputMax: Int, outputMax: Int, input: Int): Int = // min = 0
    if (inputMax == 0) 0 else (1.0 * input * outputMax / inputMax).roundToInt()

  //endregion

  override fun render(mouseX: Int, mouseY: Int, partialTicks: Float) {
    if (renderBorder) {
      rDrawOutline(absoluteBounds, borderColor)
    }
    // render scrollbar, ref: EntryListWidget.render
    if (scrollbar.visible) {
      rFillRect(scrollbar.trackAbsoluteBounds, COLOR_SCROLLBAR_BG)
      val hover = scrollbar.thumbAbsoluteBounds.contains(mouseX, mouseY) || draggingScrollbar
      rFillRect(scrollbar.thumbAbsoluteBounds, if (hover) COLOR_SCROLLBAR_HOVER_SHADOW else COLOR_SCROLLBAR_SHADOW)
      rFillRect(scrollbar.thumbAbsoluteBounds.run {
        copy(width = width - 1, height = height - 1)
      }, if (hover) COLOR_SCROLLBAR_HOVER else COLOR_SCROLLBAR)
    }

    super.render(mouseX, mouseY, partialTicks)
  }

  // scrolling logic / ui events
  private var draggingScrollbar = false
  private var draggingInitMouseY = 0
  private var draggingInitScrollbarY = 0
  override fun mouseClicked(x: Int, y: Int, button: Int): Boolean =
    super.mouseClicked(x, y, button) || if (
      button == 0 && scrollbar.visible && scrollbar.trackAbsoluteBounds.contains(x, y)) {
      if (!scrollbar.thumbAbsoluteBounds.contains(x, y)) {
        scrollbar.y = y - viewport.screenY - scrollbar.thumbHeight / 2 // e = y1 + yoffset + sh/2
      }
      draggingScrollbar = true
      draggingInitMouseY = y
      draggingInitScrollbarY = scrollbar.y
      true
    } else false

  override fun mouseReleased(x: Int, y: Int, button: Int): Boolean =
    false.also { draggingScrollbar = false }

  override fun mouseDragged(x: Double, y: Double, button: Int, dx: Double, dy: Double): Boolean =
    super.mouseDragged(x, y, button, dx, dy) || if (draggingScrollbar) {
      val shiftY = (y - draggingInitMouseY).toInt()
      scrollbar.y = draggingInitScrollbarY + shiftY
      true
    } else false

  override fun mouseScrolled(x: Int, y: Int, amount: Double): Boolean = // f = 1 or -1
    super.mouseScrolled(x, y, amount) || if (scrollbar.visible) {
      scrollY -= (amount * 20).toInt()
      true
    } else false

}